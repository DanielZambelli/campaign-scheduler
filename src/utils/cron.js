const CronJob = require('cron').CronJob

const cron = (cronStr, callback, autoStart = true) => {
  return new CronJob(cronStr, callback ,null, autoStart)
}

const cronOnce = (cronStr, callback, autoStart = true) => {
  const job = cron(cronStr, async () => {
    await callback()
    job.stop()
  }, autoStart)
  return job
}

module.exports = {cron,cronOnce}
