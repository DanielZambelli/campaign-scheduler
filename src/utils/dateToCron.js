const dateToCron = (dateString) => {
  const date = new Date(dateString)
  const seconds = date.getSeconds()
  const minutes = date.getMinutes()
  const hours = date.getHours()
  const days = date.getDate()
  const months = date.getMonth()
  const dayOfWeek = date.getDay()
  return `${seconds} ${minutes} ${hours} ${days} ${months} ${dayOfWeek}`
}

module.exports = {dateToCron}
