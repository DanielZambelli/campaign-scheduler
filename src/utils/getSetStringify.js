const getSetStringify = (field) => {
  return {
    get(){
      const value = this.getDataValue(field)
      return value ? JSON.parse(value) : value
    },
    set(value){
      if(value) value = JSON.stringify(value)
      this.setDataValue(field, value)
    }
  }
}

module.exports = {
  getSetStringify
}
