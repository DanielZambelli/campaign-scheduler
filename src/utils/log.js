const util = require('util')
const log = (object) => console.log(util.inspect(object, false, null, true))

module.exports = {log}
