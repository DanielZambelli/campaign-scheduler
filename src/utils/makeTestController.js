const { CampaignScheduler: Controller } = require('../controller/index')

const makeTestController = (id, callbacks = {}) => {

  const db = process.env.CS_DB ? JSON.parse(process.env.CS_DB) : { "dialect": "sqlite" }
  if(db.dialect==='sqlite') db.storage = `tmp/db_${id}.sqlite`

  db.schema = `test_${id}`
  db.force = true
  db.makeDestroy = true

  return new Controller({ db, callbacks }).init()
}

module.exports = {
  makeTestController
}
