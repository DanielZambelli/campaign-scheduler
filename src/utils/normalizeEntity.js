const normalizeEntity = (entity) => {
  const {createdAt,updatedAt,...entityNormalized} = entity.toJSON()
  return entityNormalized
}

module.exports = {
  normalizeEntity
}
