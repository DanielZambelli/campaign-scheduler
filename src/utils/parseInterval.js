const moment = require('moment')
const {parseIntervalOffset} = require('./parseIntervalOffset')
const {parseIntervalTime} = require('./parseIntervalTime')

const parseInterval = (interval, startDate=undefined) => {

  const {offset=null, time=null} = interval
  startDate = moment(startDate)

  //start from now or offset from previous date
  const date = moment(startDate)

  // offset
  const offsetParsed = parseIntervalOffset(offset)
  if(offsetParsed) date.add(offsetParsed)

  // specific time
  const timeParsed = parseIntervalTime(time)
  if(timeParsed) date.set(timeParsed)

  // incase now has passed the time slot offset till next day
  if(date.isSame(startDate, 'day') && startDate.isAfter(date))
    date.add({ days: 1 })

  return date.toISOString()
}

module.exports = {parseInterval}
