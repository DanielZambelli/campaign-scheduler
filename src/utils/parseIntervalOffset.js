const {getRandomBetween} = require('./getRandomBetween')

const parseIntervalOffset = (offset) => {
  if(!offset) return
  let [a,b] = offset.split('-').map(offset => {
    return offset.trim().split(' ').reduce((map,item) => {
      const value = parseInt(item.match(/\d+/g))
      const unit = item.replace(value,'')
      map[unit] = value
      return map
    }, {})
  })
  if(a && b){
    Object.keys(b).forEach(unit => {
      if(a[unit]){
        const min = Math.min(a[unit], b[unit])
        const max = Math.max(a[unit], b[unit])
        a[unit] = getRandomBetween(min,max)
      }else a[unit] = b[unit]
    })
  }
  return a
}

module.exports = {parseIntervalOffset}
