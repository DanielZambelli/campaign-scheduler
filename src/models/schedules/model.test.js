require('../../utils/test')
const TEST = 'schedules_model'

const schedule = {
  subject: 'contact#1',
  campaignId: 'campaign1'
}

describe(TEST, () => {

  beforeEach(init(TEST))

  afterEach(destroy)

  it('model', async () => {
    const res = [
      ctl.db.Schedules.rawAttributes,
      ctl.db.Schedules.associations,
      ctl.db.Schedules.underscored,
      ctl.db.Schedules.tableName,
    ]
    expect(res).toMatchSnapshot()
  })

  it('unique', async () => {
    await ctl.db.Schedules.create(schedule)
    const res = await ctl.db.Schedules.create(schedule).catch(e => e.message)
    expect(res).toEqual('Validation error')
  })

  it('unique', async () => {
    await ctl.db.Schedules.create(schedule)
    await ctl.db.Schedules.create({ ...schedule, subject: 'portal#1' })
  })

})
