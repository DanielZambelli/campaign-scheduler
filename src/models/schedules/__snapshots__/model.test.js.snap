// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedules_model model 1`] = `
[
  {
    "active": {
      "Model": [Function],
      "_modelAttribute": true,
      "defaultValue": false,
      "field": "active",
      "fieldName": "active",
      "type": BOOLEAN {},
    },
    "campaignId": {
      "Model": [Function],
      "_modelAttribute": true,
      "allowNull": false,
      "field": "campaign_id",
      "fieldName": "campaignId",
      "type": STRING {
        "_binary": undefined,
        "_length": 255,
        "options": {
          "binary": undefined,
          "length": undefined,
        },
      },
      "unique": "unique_schedule",
    },
    "createdAt": {
      "Model": [Function],
      "_autoGenerated": true,
      "_modelAttribute": true,
      "allowNull": false,
      "field": "created_at",
      "fieldName": "createdAt",
      "type": DATE {
        "_length": "",
        "options": {
          "length": undefined,
        },
      },
    },
    "id": {
      "Model": [Function],
      "_modelAttribute": true,
      "autoIncrement": true,
      "field": "id",
      "fieldName": "id",
      "primaryKey": true,
      "type": INTEGER {
        "_decimals": undefined,
        "_length": undefined,
        "_precision": undefined,
        "_scale": undefined,
        "_unsigned": undefined,
        "_zerofill": undefined,
        "options": {},
      },
    },
    "subject": {
      "Model": [Function],
      "_modelAttribute": true,
      "allowNull": false,
      "field": "subject",
      "fieldName": "subject",
      "type": STRING {
        "_binary": undefined,
        "_length": 255,
        "options": {
          "binary": undefined,
          "length": undefined,
        },
      },
      "unique": "unique_schedule",
    },
    "updatedAt": {
      "Model": [Function],
      "_autoGenerated": true,
      "_modelAttribute": true,
      "allowNull": false,
      "field": "updated_at",
      "fieldName": "updatedAt",
      "type": DATE {
        "_length": "",
        "options": {
          "length": undefined,
        },
      },
    },
  },
  {},
  true,
  "cs_schedules",
]
`;
